sympy/matrices/determinant.py

```diff
@@ -669,14 +661,17 @@ def _det(M, method="bareiss", iszerofunc=None):
                 - M[0, 1] * M[1, 0] * M[2, 2])
             return _get_intermediate_simp(_dotprodsimp)(m)
 
-    if method == "bareiss":
-        return M._eval_det_bareiss(iszerofunc=iszerofunc)
-    elif method == "berkowitz":
-        return M._eval_det_berkowitz()
-    elif method == "lu":
-        return M._eval_det_lu(iszerofunc=iszerofunc)
-    else:
-        raise MatrixError('unknown method for calculating determinant')
+    det = M.one
+    for b in M.strongly_connected_components():
+        if method == "domain-ge": # uses DomainMatrix to evalute determinant
+            det *= _det_DOM(M[b, b])
+        elif method == "bareiss":
+            det *= M[b, b]._eval_det_bareiss(iszerofunc=iszerofunc)
+        elif method == "berkowitz":
+            det *= M[b, b]._eval_det_berkowitz()
+        elif method == "lu":
+            det *= M[b, b]._eval_det_lu(iszerofunc=iszerofunc)
+    return det
```

Comment:
Currently `todok` is too slow to notice this but creating a product with repeated use of `*` is very slow compared to collecting all factors and calling `Mul` once. The `Mul.flatten()` routine is `O(N)` so repeated use of `*` is `O(N**2)`:
```python
In [19]: def prod1(n):
    ...:     d = S.One
    ...:     for _ in range(n):
    ...:         d *= Dummy()
    ...:     return d
    ...: 

In [20]: def prod2(n):
    ...:     factors = [Dummy() for _ in range(n)]
    ...:     return Mul(*factors)
    ...: 

In [21]: prod1(3)
Out[21]: Dummy₂₆⋅Dummy₂₇⋅Dummy₂₈

In [22]: prod2(3)
Out[22]: Dummy₂₉⋅Dummy₃₀⋅Dummy₃₁

In [23]: for n in [1, 10, 100]:
    ...:     %time prod1(n)
    ...: 
CPU times: user 142 µs, sys: 9 µs, total: 151 µs
Wall time: 157 µs
CPU times: user 8.05 ms, sys: 87 µs, total: 8.14 ms
Wall time: 8.31 ms
CPU times: user 347 ms, sys: 1.9 ms, total: 349 ms
Wall time: 351 ms

In [24]: for n in [1, 10, 100]:
    ...:     %time prod2(n)
    ...: 
CPU times: user 166 µs, sys: 11 µs, total: 177 µs
Wall time: 194 µs
CPU times: user 860 µs, sys: 3 µs, total: 863 µs
Wall time: 874 µs
CPU times: user 5.21 ms, sys: 88 µs, total: 5.3 ms
Wall time: 5.52 ms
```